### DISCUSSION TITLE
Future plans

### DISCUSSION THREAD

**Discussion by i-make-robots** [MarginallyClever/Makelangelo-software#414] [Ideas]
Are there any actions that should not be undoable?  New...?  FirmwareUpdate.start?

All undoable things become Actions which spawn Edits which are stored in the Undo Queue.  

The bonus here is that Actions can be moved around so users can customize their GUIs ... lol  just kidding that's a lot of work.  But in a more real sense, an Action that belongs to MarlinPlotterInterface can be borrowed by PlotterControls instead of moved completely.

**Comment by i-make-robots** [comment]
More broadly, where do we want the app to go?  Many PRs have been done lately that tweak the underlying system with little or change to the average user.  I'd like us to have consensus about what we want in the next version of the app.

Personally, I need a way to support different styles of machine like M5, M5 huge, axidraw (many versions), custom, etc.
I'd like the time estimator to be more accurate.
I'd like to LOSE voronoi and zigzag style because they are the only iterative converters and they make the code much more complicated.
Better to have them as stand alone features or use someone else's well developed voronoi generator.

**Reply by PPAC37** [reply]
> Personally, I need a way to support different styles of machine like M5, M5 huge, axidraw (many versions), custom, etc.

Yes ! as in 3d printing slicer : machine profils ( 
* profil name
* machine types ( ? ) 
  * machine / Firmware specific limitation Feedrate, Acceleration , servo delais, ...
  * gcode flavor ("marlin-polargraph" or "prolagraph-firmware" or ...), 
  * capability (servo or not, and if a servo : angle Up, angle down value settings), 
  * belt lenght (), 
  * motors distance ( X_BED_SIZE) , 
  * height reachable ( Y_BED_SIZE ) , 
* Customisable Start and End g-code ,  (to add a lot of M300 to make some musical end ;) )

* ? ...
) 
Somes are already there but not reatchable via the GUI.

But this ask some study/analysys ( posible or not, implications, implementation task complexity) 

> I'd like the time estimator to be more accurate.

Need to have all the feedrate / acceleration values ( and servo delais ) from Machine profil.

> I'd like to LOSE voronoi and zigzag style because they are the only iterative converters and they make the code much more complicated.
Better to have them as stand alone features or use someone else's well developed voronoi generator.

Plugins ?

**Reply by PPAC37** [reply]
And yes for java Action / AbstractAction for JComponants like JButtons / Jmenu Item (it's a way to have on the fly Translation language change. Icons, accelerator Key , mnemoniques , tooltips , ...) .

**Reply by PPAC37** [reply]
For the undo/redo yes too but lots of work ...

**Comment by i-make-robots** [comment]
On the fly translations changes are a tricky.  either every window has to be destroyed and recreated without losing state information OR every JLabel has to be updated.  Both are difficult.  So I went with the lazy solution, tell the user to reload the app.

**Comment by PPAC37** [comment]
Ho yes ... sorry I spoke too quickly, I forgot all the JComponent which cannot take AbstractAction in their constructor ... It requires to make special JComponents for that by inheritance ... and necessarily considering the number of JLable already present it is possibly a complicated modification to carry out well not to produce bugs.

**Comment by i-make-robots** [comment]
I just added the CI badge to the README so that new visitors can see we have CI and that it is passing.  I am looking for any ideas that will encourage more people to join the project and collaborate together.

**Comment by i-make-robots** [comment]
The more each class and action become decoupled the more I dream about a node-based editor.  The starting nodes would be art source, connected to modifiers, connected to output:gcode connected to PlotterControls (for example).  the "four tabs" branch started work on this long ago but didn't do it right.

**Reply by coliss86** [reply]
a lot of work to do that... It is the way osgi is used for. I think it is a technical challenge, but to few benefit for users.

**Comment by i-make-robots** [comment]
@coliss86 @PPAC37 if the x-platform installer is ready...
- [x] include installer in the nightly builds
- [x] bump version to 7.32
- [x] update README.md
- [ ] update Dozuki instructions
- [ ] prep announce for FF with new features list
- [ ] new features video for youtube?
- [ ] make example picture into embedded youtube example video.
- [ ] release!

**Reply by coliss86** [reply]
oh yeah !!!

**Comment by i-make-robots** [comment]
- [ ] make example picture into embedded youtube example video.
